hs-bindgen/src-internal/HsBindgen/BindingSpec/Gen.hs

```diff
@@ -240,23 +234,35 @@ genBindingSpec'
 -- if a name is detected as anonymous, which we cannot reliably do anymore. I'm
 -- wondering if we can get rid of this source altogether, and simply record the
 -- C name as part of the DeclInfo.
-getCQualName :: C.DeclInfo -> C.NameKind -> C.QualName
-getCQualName declInfo cNameKind =
-    case declInfo.declId of
-      C.DeclIdNamed named ->
-        case named.origin of
-          C.NameOriginInSource ->
-            C.QualName cName cNameKind
-          C.NameOriginGenerated{} ->
-            let cName' = fromMaybe cName (listToMaybe (C.declAliases declInfo))
-            in  C.QualName cName' C.NameKindOrdinary
-          C.NameOriginRenamedFrom fromCName ->
-            C.QualName fromCName cNameKind
-      C.DeclIdBuiltin builtin ->
-        C.QualName builtin.name cNameKind
+getCQualName :: C.FinalDeclId -> Maybe C.QualName
+getCQualName declId =
+    case declId.origDeclId of
+      C.OrigDeclId orig ->
+        Just $ auxOrig orig
+      C.AuxForDecl _parent ->
+        -- TODO <https://github.com/well-typed/hs-bindgen/issues/1379>
+        Nothing
   where
-    cName :: C.Name
-    cName = C.declIdName declInfo.declId
+    auxOrig :: PrelimDeclId -> QualName
+    auxOrig = \case
+        PrelimDeclIdNamed cName kind ->
+          C.QualName cName kind
+        PrelimDeclIdAnon _anonId kind ->
+          -- Anonymous declarations can only arise in very specific situations:
+          --
+          -- * Anonymous declarations without any use sites are removed
+          -- * Anonymous declarations inside typedefs are given a name, but their
+          --   /original/ declaration is set to be the typedef instead.
+          --
+          -- However, anonymous structs inside other structs or unions will
+          -- remain anonymous. For these we instead use the name assigned by
+          -- the @NameAnon@ pass.
+          --
+          -- TODO <https://github.com/well-typed/hs-bindgen/issues/844>
+          -- This is WIP.
+          C.QualName ("@" <> declIdName declId) kind
```

Comment:
This isn't _quite_ right I think. We should be able to rename declarations further, if needed for other reasons. Ideally we'd have the `DeclId NameAnon` name here in the `origDeclId`, so that we have the name assigned by `NameAnon` always available (which is meant not to be configurable precisely so that we can refer to it).